Allow overriding real_ip_recursive

tomaspinho opened the pull request:
<!--- Provide a general summary of your changes in the Title above --->
<!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). --->

## What this PR does / why we need it:
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Allows control over `real_ip_recursive`, fixing problem shown in issues like: https://github.com/kubernetes/ingress-nginx/issues/4073

We are using this to fix a `X-Forwarded-For` spoofing issue we found while using AWS ALB as a service load balancer. Turns out they respect upstream `X-Forwarded-For` and don't allow us to strip/replace them, causing NGINX Ingress Controller to take the first value of the header as the defacto client IP, which can be spoofed.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):

fixes #
-->

https://github.com/kubernetes/ingress-nginx/issues/4073

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

longwuyuan commented:
@tomaspinho thanks for contributing to this long pending issue. Helps a lot.

Any chance that a e2e test can be added that simulates the complete event sequence from a client spoofing the ip with X-Forward-For header and the controller picking one the last "real" client_ip_address.

tomaspinho commented:
It's my first time contributing to this project, so I have to learn how to run and write e2e tests.

I hope to have some time this weekend to take a look at this. Thanks @longwuyuan

tomaspinho commented:
/remove-lifecycle stale

tomaspinho commented:
@longwuyuan looks like this one was never merged, but it fixes a real issue. Would it be possible to consider reopening and performing a final review?

tomaspinho commented:
Also, cc @rikatz @strongjz

tomaspinho commented:
We're keeping a custom version of NGINX Ingress Controller's `nginx.conf` in order to work around this issue, but it still very much affects us and I would love to see this merged :D

ivan909020 commented:
Any news to merge this fix?

michalziobro commented:
looks like this PR was abandoned but it addresses a vital issue. Are there any chances to merge?

tomaspinho commented:
Hey @rikatz @strongjz! :smile: 

Just rebased this PR and fixed the linting/formatting according to upstream. Wondering if there's anything you need from me in order to merge? Thanks!

rikatz commented on internal/ingress/controller/config/config.go:
```diff
@@ -790,6 +795,7 @@ func NewDefault() Configuration {
 		ErrorLogLevel:                    errorLevel,
 		UseForwardedHeaders:              false,
 		EnableRealIP:                     false,
+		EnableRealIPRecursive:            true,
```
shouldn't this be disabled by default?

rikatz replied:
there is a "problem" with this approach of a flag defaulted to true: if you set it as false on your configmap, it may or may not represent as empty (depending on the marshaller, eg using omitempty on json flag) so because it is empty it will be defaulted to...true :) 

To confirm this behavior, can you add some e2e test with both EnableRealIPRecursive = true or false on the configmap, and check if the template is rendered correctly?

tomaspinho replied:
Hi @rikatz ,

I've set the default to `false`. My initial reasoning for it to be `true` was that I think the more secure option should be enabled by default if it doesn't otherwise impact any functionality. Because the setting itself is dependent on `EnableRealIp` being set to `true`, the more secure configuration would be the default when `EnableRealIp` would be used. This is a taste thing, and not a deal breaker for getting this merged.

rikatz commented on test/e2e/settings/enable_real_ip_recursive.go:
```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
nit: fix copyright year

strongjz commented:
/retest

strongjz commented:
/triage accepted
/kind feature
/priority backlog

strongjz commented:
/ok-to-test

daanwa commented:
Any progress regarding this merge? We are having same issue

dkulchinsky commented:
@strongjz anything blocking this from getting accepted?
